### DISCUSSION TITLE
Issue with generics

### DISCUSSION THREAD

**Discussion by kostysh** [sinclairzx81/typebox#351] [Q&A] [answered]
Hi, I am trying to implement a generic type that accepts a custom schema as an argument and builds a new schema that incorporates this custom schema. I have some issues that I mention in the code below.
What I am doing wrong? How to create and use generics in the right way?

And another couple of questions: 
- What is the difference between `Value.Convert` and `Value.Cast`?
- Is it possible to assert and Cast/Convert type at once as it is implemented in other libraries (where you can call the `parse` or `validate` method of schema and this method will throw in case of broken input data type)?

```typescript
import { Type, Static, TSchema } from '@sinclair/typebox';
import { Value } from '@sinclair/typebox/value';

export const GenericQuerySchema = Type.Object({});

export type GenericQuery = Static<typeof GenericQuerySchema>;

const BaseMessageSchema = Type.Object({
  id: Type.Optional(Type.String()),
});

const createOptsSchema = <T extends TSchema>(querySchema: T) =>
  Type.Composite([
    BaseMessageSchema,
    Type.Object({
      querySchema: Type.Any(),
      query: querySchema,
    }),
  ]);

type OPTS<T extends TSchema> =
  Static<ReturnType<typeof createOptsSchema<T>>>;

const test = <T extends TSchema>(options: OPTS<T>) => {
  const optsSchema = createOptsSchema<T>(options.querySchema);
  // ------------------------------------------^ Property 'querySchema' does not exist on type 'Evaluate<Evalu...
  const errors = Value.Errors(optsSchema, options);
  console.log('Errors: -->', errors.First());
  options = Value.Cast(optsSchema, options);
  console.log('Options: -->', options.id, options.query);
   // ------------------------------------------^ Property 'query' does not exist on type 'Evaluate<Evalu...
};

const CustomQuerySchema = Type.Object({
  hello: Type.String(),
});

test<typeof CustomQuerySchema>({
  id: '123',
  query: {
    hello: 'qwerty',
  },
  querySchema: CustomQuerySchema,
});
```

> `@sinclair/typebox` v0.26.2
> `typescript` v5.0.2

**Comment by sinclairzx81** [comment, marked as answer]
@kostysh Hi. 

Instantiation Expressions mixed with TypeBox types can be quite tricky to get right, but the following should implement something close to what you're trying to achieve.

[TypeScript Link Here](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgFQJ5gKYBo4GUYCGMwAxjsriQBYYgFwC+cAZlBCHAOQACAzsADsSAGwLAoAehjoMAIwgAPTgFgAUGokS4AWl179Bw0eMnTZ-Rq0BxDAIxRScAIoBXe6jgB5MMQgDeljrmwSGhoWpqGAqQsHDSmHBePsB+vAA8yHBRMLYAJrwolDR0AHxwALx4hMQkaQBKGDAuUAJomGnxGBDMSb7+GSWDkdHQ8CSp8N59BZUZWQo5AvmF1LQEJQAUAI5uUKgAXCgAlBVlbRgAdADC7JD8ORsA2mqJ5xeesgBWGCQwG0jAXKHN74BwCADmGxODCOWBeKBk7y+Pz+CHhiR27iKa2BiIAggJUFC4apEhjdqh4TC1ABdI4RVSaIJhFmslmBACyjSoEHygTZAsFpgZzBcQj6cQwvBgc2yeQKUxS-Qoq1KmwgyVSwJOaNJcHG-ngANyOExe0YFTgGumcCZZo8wAKnDFAGsBBAAO4CThwAgFFXFeiOuAANwIe0E4LUDDUovFSsl0oA+gbpVAxHZcrKFvLekr0p1uihBhtrfntYh4amjXBAaaKRbKmXUratPba06kARDgIXCBZPYcLIe32B1BGCpVDH1IytEL5wvtIEAKIKAjgYQYfmLneshnVlCWt4fb6-f7w7sIzAXAByo-sxPhw6vlzv-Yf9Knn7UOWlaUVqQdDIRbICWSBMnKLQEMIcC5BgmBLLY8CdPCgKHJwACMABMADMnAkuSWKqpeCAMARcD2ocupkr6hwYeRiTPhhVLRp+v4wCmEzpoIGC5OeepoVw2F4eR9rYnQVFkfClGVnqiSXvR6JwExLFfkAA)

```typescript
import { Type, Static, TSchema } from '@sinclair/typebox'

// ----------------------------------------------------------------------
// Generic Query Options
// ----------------------------------------------------------------------

export type  Options<T extends TSchema> = Static<ReturnType<typeof Options<T>>>
export const Options = <T extends TSchema>(query: T) => Type.Composite([
  Type.Object({ id: Type.String() }),
  Type.Object({
    querySchema: Type.Any(),
    query
  })
])

// ----------------------------------------------------------------------
// Methods
// ----------------------------------------------------------------------

function test<T extends Options<TSchema>>(options: T) {
  const { id, query } = options // query is 'unknown' as TSchema is varying
}
function test_constrained<T extends Options<typeof T>>(options: T) {
  const { id, query } = options // query is '{ a: number, b: number }'
}

// ----------------------------------------------------------------------
// Example
// ----------------------------------------------------------------------

const T = Type.Object({
  a: Type.Number(),
  b: Type.Number()
})

test<Options<typeof T>>({ // external dependent type
  id: '123',
  querySchema: {},
  query: {
    a: 1,
    b: 1
  }
})
test_constrained({
  id: '123',
  querySchema: {},
  query: {
    a: 1,
    b: 1
  }
})
```
Hope this helps!
S

**Reply by kostysh** [reply]
Perfect! Thank you very much for so a clear explanation! I have spent so much time solving this issue with the `zod` and `yup` libraries and without any success, even with community help. The `typebox` is awesome!

- What is the difference between `Value.Convert` and `Value.Cast`?
- Is it possible to assert and Cast/Convert type at once as it is implemented in other libraries (where you can call the parse or validate method of schema and this method will throw in case of broken input data type)?
- How to create a schema for validating properties which are types of `typebox` schemes? Like in my example where `querySchema` should be an instance of `Type.Object(...)`.

Sorry for disturb...

**Comment by sinclairzx81** [comment]
@kostysh Hi

> Perfect! Thank you very much for so a clear explanation! I have spent so much time solving this issue with the zod and yup libraries and without any success, even with community help. The typebox is awesome!

You're welcome, and thanks!

> What is the difference between Value.Convert and Value.Cast?

### Cast

The Cast function is similar to the Create function, but is different in that it accepts a "Partial" value and will populate missing values with defaults while retaining valid values. Cast was written specifically with data migrations in mind. Scenarios where you might use Cast is when you change a schematic and need to update existing values to match the new schema / type.

```typescript
const T = Type.Object({
  x: Type.Number(),
  y: Type.Number(),
  z: Type.Number(),
})

const V = Value.Cast(T, { x: 42 }) // const V = { x: 42, y: 0, z: 0 } - retain { x: 42 } add { y: 0, z: 0 }
```

### Convert

Convert is a new function added on 0.26.0. It does value coercion to try and align a value to its target type. It's important to note that conversion can fail, and because of this, the Convert function will return `unknown` (even if the conversion was successful). It is intended to be used in tandem with other functions (like Check and Cast)

```typescript
import { Type } from '@sinclair/typebox'
import { Value } from '@sinclair/typebox/value'

const T = Type.Object({ x: Type.Number() }) // target type { x: number }

const C1 = Value.Convert(T, { x: '42' })    // const C1: unknown = { x: 42 }      - x could be converted from '42' to 42
const C2 = Value.Convert(T, { x: 'hello' }) // const C2: unknown = { x: 'hello' } - x could not be converted

const R1 = Value.Check(T, C1)              // const R1 = true  - value conversion resulted in correct value
const R2 = Value.Check(T, C2)              // const R2 = false - value conversion resulted in incorrect value
```
> Is it possible to assert and Cast/Convert type at once as it is implemented in other libraries (where you can call the parse or validate method of schema and this method will throw in case of broken input data type)?

You can use Convert and Cast together in the following way
```typescript
const T = Type.Object({ x: Type.Number(), y: Type.Number() })

const convert = Value.Convert(T, { x: '42' }) // const convert = { x: 42 }

const casted = Value.Cast(T, convert)         // const casted  = { x: 42, y: 0 }
```
> How to create a schema for validating properties which are types of typebox schemes? Like in my example where querySchema should be an instance of Type.Object(...).

The simplest way to approach this is with a custom type. Below is a revised version of the `Options<T>` type.

[TypeScript Link Here](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgFQJ5gKYGVUGcYYhwC+cAZlBEQOQACuwAdgMYA2AhsFAPQzoYAjCAA9uuPARDUAsAChQkWIhT8ANCizMAFoXbqsMdjGDN1aTAHEAruygATEuUo16TNpx59MQ4TNlzubjgAWlCw8IjIqOiY2LjwgKCLDEYMKBM4AEUrNNQ4AHkwYwhGXDgAag1tXThmK3wqOC8MRJD49o7OzrlmEvwqnRB2OABeFUwcfEIAOnMMAB5kTUH2AD4ACmol6qHqdXWAfXUAN3ZWHIBKUdXxjGtbO1nl3XXT84wLq7k5DGFFeGacAKRWAfUWcF+BEYdjK2xWNzGBiMJnmACUMDArFBGHN5s0IGRgcVSotVmSfn9oPBeqV4IViWUxuDISkYQNdBsAI45KCoABcKCuIxuc2mAGEqJAGAR1gBtORA0X5AQAKwwzBg6yQwDsAtFBnSjAA5usrsQLqoFbdpsq1RqtVagdzcs8hnrXexTZbZECnTzUFbzXIALoXb6yQJtLrRmPR1oAWQxWggMNasfTGdi4bIVhYxKaGHwzOEULZ9NBJLhHI2EBBfT1VwQVpp-W1dnUzt5Hf9HscY1rDLkxDkObzFYL+AOLZgUE4qTsxdLZXLYPxhOQZPWA4ruAbiGbfXgbe7uRPvN7pH7ddKcCH4cjmcfT+CrQAosJ2OBWC0I0Fn--Y3DacUFGa1bXVTUmx9OB2D1fhpgAOSsEABDSL0rQEODMEQ5DUKgU0hzDfxZAIIsVxJNcUE3KCgR1AVqAARgAJgAZj2K1O1QD0sIwaYDSYE0LQ4-0BRo31YLgBjvV9OBMMkwNCLkUiYCnQ9ZyYDA7AdaC6LgRjWPY6DOO461+ONdCjJE-doKBCSpMdWSBQYhTZHNIA)

```typescript
import { TypeSystem } from '@sinclair/typebox/system'
import { Type, TSchema, Static, TypeGuard } from '@sinclair/typebox'

// ----------------------------------------------------------------------
// Generic Query Options + TSchema custom type
// ----------------------------------------------------------------------
const TSchema = TypeSystem.Type<TSchema>('TSchema', (_, value) => TypeGuard.TSchema(value)) 

export type  Options<T extends TSchema> = Static<ReturnType<typeof Options<T>>>
export const Options = <T extends TSchema>(query: T) => Type.Composite([
  Type.Object({ id: Type.String() }),
  Type.Object({
    querySchema: TSchema(),
    query
  })
])

// ----------------------------------------------------------------------
// Methods
// ----------------------------------------------------------------------

function test<T extends Options<TSchema>>(options: T) {
  const { id, query, querySchema } = options
}
function test_constrained<T extends Options<typeof T>>(options: T) {
  const { id, query, querySchema } = options 
}

// ----------------------------------------------------------------------
// Example
// ----------------------------------------------------------------------

const T = Type.Object({
  a: Type.Number(),
  b: Type.Number()
})

test<Options<typeof T>>({
  id: '123',
  querySchema: Type.String(),
  query: {
    a: 1,
    b: 1
  }
})
test_constrained({
  id: '123',
  querySchema: Type.String(),
  query: {
    a: 1,
    b: 1
  }
})
```
There is also a way to express a runtime `TSchema` using TypeBox types (without custom types), but unfortunately, this approach will make TS inference explode. The following will runtime assert ok, however, I wouldn't recommend this due to the inference issues you'll run into.

```typescript
const Object = <T extends TSchema>(TSchema: T) => Type.Object({ type: Type.Literal('object'), required: Type.Optional(Type.Array(Type.String())), properties: Type.Record(Type.String(), TSchema) })
const Array = <T extends TSchema>(TSchema: T) => Type.Array(TSchema)
const String = Type.Object({ type: Type.Literal('string') })
const Number = Type.Object({ type: Type.Literal('number') })
const Boolean = Type.Object({ type: Type.Literal('boolean') })

const TSchema = Type.Recursive(TSchema => Type.Union([ // union all types of TSchema
    Object(TSchema),
    Array(TSchema),
    String,
    Number,
    Boolean,
]))
```
Hope this helps
S

**Reply by kostysh** [reply]
Thank you! That is it!

**Comment by kostysh** [comment]
@sinclairzx81 Tried to dive deeper into the case discussed above and stuck with the following issue.
Because this code is dedicated to the library It is not possible to use a `constrained` version.

```typescript
function test_constrained<T extends Options<typeof T>>(options: T) {
  const { id, query, querySchema } = options 
}
```

this `T` schema is not known at this moment. This `T` schema is a custom schema that will be created by the library user (developer) later.
So, the only possible option is to use an unconstrained version where is `query` is `unknown`

```typescript
function test<T extends Options<TSchema>>(options: T) {
  const { id, query, querySchema } = options;
  // ...
  return options;
}
```

Let's look into the example. Here I have created a `Message` type that is a generic type that includes `query` as a property. This message is constructed inside the example function `test`. Here is the issue. The internal `Message` type that will be returned from the example function will be different from the external `Message` (I have commented on it in the code).
Is it possible somehow to return the right `Message` type? not with `unknown` `query`.
I have tried to change the example definition in the following way: `function test<T extends TSchema, O extends Options<T>>(options: O): Message<T> ` but it is not working, this method breaks `options`.

[Link to the playground](https://www.typescriptlang.org/play?module=1&allowSyntheticDefaultImports=true#code/JYWwDg9gTgLgBAbzgFQJ5gKYGVUGcYYhwC+cAZlBEQOQACuwAdgMYA2AhsFAPQzoYAjCAA9uuPARDUAsAChQkWIhT8ANCizMAFoXbqsMdjGDN1aTAHEAruygATEuUo16TNpx59MQ4TNlzubjgAWlCw8IjIqOiY2LjwgKCLDEYMKBM4AEUrNNQ4AHkwYwhGXDgAag1tXThmK3wqOC8MRJD49o7OzrlmEvwqnRB2OABeFUwcfEIAOnMMAB5kTUH2AD4ACmol6qHqdXWAfXUAN3ZWHIBKUdXxjGtbO1nl3XXT84wLq7k5DGFFeGacAKRWAfUWcF+BEYdjK2xWNzGBiMJnmACUMDArFBGHN5s0IGRgcVSotVmSfn9oPBeqV4IViWUxuDISkYQNdBsAI45KCoABcKCuIxuc2m+QEACsMMwYOsEHIgdzcs8hgK4S8LqoFXAlby5MQLt9ZL9-k1+ECALIYXC4dgAcwWyAhwihbPVQwRcCRxmYaIxWJx-Dx-AJcCtNvtjrJqwpppp-XDtodozgzJdrNhKrW611-MF11uYsl0tl8tkQOAdjV-GmBnSjDt6012tz+sN-lkgTaXR7vZ7rStMC0EBhrT744nsSNZCsLGJTWtMDTrrK9NBJPdaw2EBBfTVVzLQPj8CQufUuazjjGO4ZRqPfXgIGtSYwAsTkeDmFDF52axTh6BOBKwFagAEYACYAGY9m1RUeVQbViAAbjvOAoH9bE4CfCMHRQ2RiCNLtJ2IkjglaABRYR2HAVgWk7IJSMYvsjWPLJ4JTUVxSlGU5W1dhq0waYADkrBAAQ0ibLVyzgAQBIwYTRPEqAmzbFiHyw3A7QASUYAhsTOFMCHweY1zBfFCWyXIyV46SfxWAVLN5KS4NyAUAKBfi4FA5zANkrzENUjtuAAKmC1oAAMQE0nS9MYM5wrNTA5EBKLtN0tI4tYf8W3ggVZwAa0YCAAHdGDwisqzgfB6ztPCCPLYLuCNFLNMo2KDLGd8HU-DBQ0c1BVjwgJQuS81UrajKOsQHLXOm6SPIFRhFLScrfMW5aoFW5DtWAqqYBquqFUaoA)

```typescript
export type  Message<T extends TSchema> = Static<ReturnType<typeof Message<T>>>
export const Message = <T extends TSchema>(query: T) => Type.Object({
  id: Type.String(),
  query
})

function test<T extends Options<TSchema>>(options: T) {
  const { query, querySchema } = options

  const message: Message<typeof querySchema> = {
    id: '123',
    query
  };

  return message;
}

const msgInternal = test<Options<typeof Query>>({
  querySchema: Query,
  query: {
    a: 1,
    b: 1
  }
})

/**
// `msgInternal` type
type msgInternal = {
  query: unknown;
  id: string;
}
 */

type msgExternal = Message<typeof Query>;

/**
type msgExternal = {
  query: {
    a: number;
    b: number;
  };
  id: string;
}
 */
```
